Update the data structure of problem instances

MehmetC-1QBit opened the pull request:
### Description

> Describe the purpose of this PR along with any background information and the impacts of the proposed change. For the benefit of the community, please do not assume prior context.

The data structure of the instance files need to be updated to address end user expectations.
The change is not vital for the solvers but required for further investigation during the data analysis.

Here is the summary of the change:
1. The old instance file in `examples/test_instances/` is replaced with new instance
2. The old instance file in `examples/tuning_instances/` is replaced with new instance 
3.  Based on #119, 
    - `ProblemInstance` is updated (including docstring) in `ccvm_simulators/problem_classes/boxqp/problem_instance.py`
    - `Solution` is updated (including docstring) in `ccvm_simulators/solution.py`
4. `CHANGELOG.md` is updated
5. `README.md` is fixed to reflect the change from previous PR
6. `ccvm_simulators/problem_classes/README.md` is updated
7. Also `examples/README.md` is updated similar to item 6 (Reviewer's attention is required the instance info is redundant with item 6)
8. The modules `dl_solver.py`, `mf_solver.py`, and `langevin_solver.py` in `ccvm_simulators/solvers/` are updated

### References

Closes #119 

### Testing

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [x] I have added documentation for new/changed functionality in this PR
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `main`

KatieO4t1qbit commented on ccvm_simulators/problem_classes/boxqp/problem_instance.py:
```diff
@@ -143,10 +157,18 @@ def load_instance(
                         rval_q[idx, j] = -torch.Tensor([float(value)])
+
+                # Read the last line as an additional information
+                last_raw_data_line = lines[-1].split("\n")[0].split(file_delimiter)
```
We'll need to be more explicit, since this line is optional. Otherwise, if the user leaves out the final line then optional_data will become populated with the last line of q_matrix.
```suggestion
                last_raw_data_line = lines[problem_size + 3].split("\n")[0].split(file_delimiter)
```

MehmetC-1QBit replied:
Any suggestion for this?

KatieO4t1qbit replied:
I think you may have forgotten to push this change?

KatieO4t1qbit reviewed (changes requested):
Thanks for this change, Mehmet!
Let's ensure that the tests pass as part of this change by running `pytest ccvm_simulators/problem_classes/boxqp/tests/test_problem_instance.py` (assumes you are in the top-level directory)

FarhadK-1QBit commented on ccvm_simulators/problem_classes/boxqp/problem_instance.py:
```diff
@@ -143,10 +157,18 @@ def load_instance(
+                optional_info = []
+                for v in last_raw_data_line:
+                    if not v == "":
+                        optional_info.append(float(v))
```
Give this a descriptive name such as `solution_vector` instead of `optional info`.

MehmetC-1QBit replied:
done

FarhadK-1QBit commented on ccvm_simulators/solution.py:
```diff
@@ -18,6 +18,9 @@ class Solution:
         optimal_value (float): The optimal objective value for the given problem instance.
+        best_value (float): The best objective value for the given problem instance.
+        num_frac_values (int): Number of fractional values in the solutions of given problem instance.
+        optional_info (list): Optional information in a given problem instance.
```
Similar change here. Describe as the solution to the problem instance instead of optional information.

MehmetC-1QBit replied:
done

FarhadK-1QBit commented on ccvm_simulators/solvers/dl_solver.py:
```diff
@@ -433,6 +435,9 @@ def _solve(
             optimal_value=instance.optimal_sol,
+            best_value=instance.best_sol,
+            num_frac_values=instance.num_frac_values,
+            optional_info=instance.optional_info,
```
Apply the previous comments here as well.

FarhadK-1QBit commented on ccvm_simulators/solvers/langevin_solver.py:
```diff
@@ -380,6 +383,9 @@ def _solve(
             optimal_value=instance.optimal_sol,
+            best_value=instance.best_sol,
+            num_frac_values=instance.num_frac_values,
+            optional_info=instance.optional_info,
```
And in this file.

MehmetC-1QBit replied:
done

FarhadK-1QBit commented on examples/README.md:
```diff
@@ -20,10 +20,18 @@ The following folders contain our example problem instances:
+
+The rest of the lines before the last line hold the elements of the `Q` matrix.
+
+The final line contains list of additional (optional) information in the problem instance.
```
Say "the final line contains the vector of the solution to the problem instance".

FarhadK-1QBit reviewed (changes requested):
Thanks for implementing this change, Mehmet! 
I only have some minor comments and changes regarding the last line of the problem instance.

MehmetC-1QBit commented on CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 - Updated internal data handling that have improved performance and reduced memory usage.
+- Updated data structure of the problem instance
```
I added the name of the additional parameters in the changelog.md.

MehmetC-1QBit commented on ccvm_simulators/problem_classes/boxqp/problem_instance.py:
```diff
@@ -126,11 +135,16 @@ def load_instance(
+                sol_time_bfgs = float(instance_info[5])
+                num_frac_values = int(
+                    instance_info[7]
+                )  # seed=int(instance_info[6]) # discarded
```
Well! Based on our discussion I just disregarded it because we have already `seed` info from the instance file name.

MehmetC-1QBit commented on ccvm_simulators/problem_classes/boxqp/problem_instance.py:
```diff
@@ -44,19 +44,28 @@ def __init__(
+            num_frac_values (int): number of fractional values in the solution
             q_matrix (torch.tensor): Q matrix of the QP problem. Defaults to None.
             v_vector (torch.tensor): V vector of the QP problem. Defaults to None.
+            optional_info (list): additional information in the problem instance
```
- @KatieO4t1qbit  I guess @FarhadK-1QBit 's answer is good enough to move forward:)
- Updated docstring of the field `optional_info` based on @FarhadK-1QBit 's comment.

MehmetC-1QBit commented:
> Thanks for this change, Mehmet! Let's ensure that the relevant tests pass as part of this change by running `pytest ccvm_simulators/problem_classes/boxqp/tests/test_problem_instance.py` (assumes you are in the top-level directory)

I updated test modules and replaced old instance files.

MehmetC-1QBit commented on ccvm_simulators/problem_classes/README.md:
```diff
@@ -31,12 +31,18 @@ independet of the particular random instances generated.
-The rest of the lines hold the elements of the `Q` matrix.
+The rest of the lines before the last line hold the elements of the `Q` matrix.
+
+The final line contains list of additional (optional) information in the problem instance.
```
Done

KatieO4t1qbit commented on problem_variables.pt:
What is this file?

MehmetC-1QBit replied:
I remember such file name was in one of test cases. It was unintentionally created during debugging session. 
Now deleted.

MehmetC-1QBit commented on CHANGELOG.md:
```diff
@@ -18,6 +18,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Updated internal data handling that have improved performance and reduced memory usage.
+- Updated data structure of the problem instance file:
+	-- To extract all the values in the first line of the instance, additional fields i.e. (`best_sol`, `sol_time_bfgs`, `num_frac_values`) are added.
+	--  The field`solution_vector` is added to extract vector of the solution to the problem instance at the end of the new instance file compared the old data structure
```
To answer your questions, let's keep in mind that we are extending the extraction of input data from a new problem instance. These fields are added because end users will have an option for further investigation.
Anyway, thank you for the suggestion: I modified your answer and rewrite both terms

MehmetC-1QBit replied:
Thanks for the catch. I updated that line based on your suggestion. Please check out the commit

MehmetC-1QBit commented on ccvm_simulators/solution.py:
```diff
@@ -18,6 +18,9 @@ class Solution:
         optimal_value (float): The optimal objective value for the given problem instance.
+        best_value (float): The best objective value for the given problem instance.
+        num_frac_values (int): Number of fractional values in the solutions of given problem instance.
+        solution_vector (list): Optional information in a given problem instance.
```
Thanks for the modification. I put in the doctstring except `Optional.`

MehmetC-1QBit replied:
Thanks for the suggestion. I updated the line 23 but this time I excluded "Optional." Because the new input file has it.

KatieO4t1qbit commented on ccvm_simulators/problem_classes/boxqp/tests/test_problem_instance.py:
```diff
@@ -97,19 +97,48 @@ def test_load_instance_cpu(self):
+        self.num_frac_values = num_frac_values
+        self.q_matrix = rval_q
+        self.v_vector = rval_v
+        self.optional_info = optional_info
```
I think this should be solution_vector now?

MehmetC-1QBit replied:
Thanks Katie, I removed the lines between 101-113 which was actually a commented segment that I copied from problem_instance.py but forgot to delete. Now it is done.

KatieO4t1qbit reviewed (approved):
Thanks Mehmet! Looking good 🔥 
I left a couple comments, please address them before merging. Thanks!

KatieO4t1qbit commented on ccvm_simulators/problem_classes/boxqp/problem_instance.py:
```diff
@@ -143,10 +157,18 @@ def load_instance(
                         rval_q[idx, j] = -torch.Tensor([float(value)])
+
+                # Read the last line as an additional information
+                last_raw_data_line = lines[problem_size + 2].split("\n")[0].split(file_delimiter)
+                solution_vector = []
+                for v in last_raw_data_line:
+                    if not v == "":
+                        solution_vector.append(float(v))
+                        
```
Is the solution vector is required before parsing? If it is optional we will need to wrap this in a try block to allow the problem to be read without error in the case that the solution vector is not given:
```suggestion
                 # Read the last line as an additional information
                solution_vector = []
                try:
                    last_raw_data_line = lines[problem_size + 2].split("\n")[0].split(file_delimiter)
                    for v in last_raw_data_line:
                        if not v == "":
                            solution_vector.append(float(v))
                except IndexError:
                    # solution_vector was not supplied
                    pass
```

KatieO4t1qbit reviewed (approved):
Looks great! Thank you, Mehmet 🙌

FarhadK-1QBit reviewed (approved):
This Look great, Mehmet! Thank you.
